### PR TITLE
OP_CODES for stack

### LINKED ISSUE #107: DeFi OP_CODES implementation mapping
> This tracks all the OP_CODES class and their categories to be implemented in jellyfish-transaction. DeFi blockchain custom transaction will be tracked in another issue.

```
packages/jellyfish-transaction/src/script/*
```

## `data.ts`
- [x] 0x01-0x4b @fuxingloh #106
- [x] OP_PUSHDATA1 @fuxingloh #106
- [x] OP_PUSHDATA2 @fuxingloh #106
- [x] OP_PUSHDATA4 @fuxingloh #106

## `constants.ts`
- [x] OP_0 @fuxingloh #106
- [x] OP_FALSE @fuxingloh #106
- [x] OP_1NEGATE @jingyi2811 #198
- [x] OP_RESERVED @jingyi2811 #217 
- [x] OP_1 @jingyi2811 #200
- [x] OP_TRUE @jingyi2811 #200
- [x] OP_2 @jingyi2811 #200
- [x] OP_3 @jingyi2811 #200
- [x] OP_4 @jingyi2811 #200
- [x] OP_5 @jingyi2811 #200
- [x] OP_6 @jingyi2811 #200
- [x] OP_7 @jingyi2811 #200
- [x] OP_8 @jingyi2811 #200
- [x] OP_9 @jingyi2811 #200
- [x] OP_10 @jingyi2811 #200
- [x] OP_11 @jingyi2811 #200
- [x] OP_12 @jingyi2811 #200
- [x] OP_13 @jingyi2811 #200
- [x] OP_14 @jingyi2811 #200
- [x] OP_15 @jingyi2811 #200
- [x] OP_16 @jingyi2811 #200

## `control.ts` 
- [x] OP_NOP = 0x61 @siradji #333 
- [x] OP_VER = 0x62 @siradji #333 
- [x] OP_IF = 0x63 @siradji #333 
- [x] OP_NOTIF = 0x64 @siradji #333 
- [x] OP_VERIF = 0x65 @siradji #333 
- [x] OP_VERNOTIF = 0x66 @siradji #333 
- [x] OP_ELSE = 0x67 @siradji #333 
- [x] OP_ENDIF = 0x68 @siradji #333 
- [x] OP_VERIFY = 0x69 @siradji #333 
- [x] OP_RETURN @fuxingloh #106

## `stack.ts`
- [x] OP_TOALTSTACK = 0x6b @aikchun #389 
- [x] OP_FROMALTSTACK = 0x6c @aikchun #389 
- [x] OP_2DROP = 0x6d @aikchun #389 
- [x] OP_2DUP = 0x6e @aikchun #389 
- [x] OP_3DUP = 0x6f @aikchun #389 
- [x] OP_2OVER = 0x70 @aikchun #389 
- [x] OP_2ROT = 0x71 @aikchun #389 
- [x] OP_2SWAP = 0x72 @aikchun #389 
- [x] OP_IFDUP = 0x73 @aikchun #389 
- [x] OP_DEPTH = 0x74 @aikchun #389 
- [x] OP_DROP = 0x75 @aikchun #389 
- [x] OP_DUP @fuxingloh #106
- [x] OP_NIP = 0x77 @aikchun #389 
- [x] OP_OVER = 0x78 @aikchun #389 
- [x] OP_PICK = 0x79 @aikchun #389 
- [x] OP_ROLL = 0x7a @aikchun #389 
- [x] OP_ROT = 0x7b @aikchun #389 
- [x] OP_SWAP = 0x7c @aikchun #389 
- [x] OP_TUCK = 0x7d @aikchun #389 

## `splice.ts`
- [x] OP_CAT = 0x7e, @jingyi2811 #510 
- [x] OP_SUBSTR = 0x7f, @jingyi2811 #510 
- [x] OP_LEFT = 0x80, @jingyi2811 #510 
- [x] OP_RIGHT = 0x81, @jingyi2811 #510 
- [x] OP_SIZE = 0x82, @jingyi2811 #510 

## `bitwise.ts` 
- [x] OP_INVERT = 0x83 @surangap #422
- [x] OP_AND = 0x84 @surangap #422
- [x] OP_OR = 0x85 @surangap #422
- [x] OP_XOR = 0x86 @surangap #422
- [x] OP_EQUAL @fuxingloh #106
- [x] OP_EQUALVERIFY @fuxingloh #106
- [x] OP_RESERVED1 = 0x89 @surangap #422
- [x] OP_RESERVED2 = 0x8a @surangap #422

## `arithmetic.ts`
- [x] OP_1ADD = 0x8b @jingyi2811 #512
- [x] OP_1SUB = 0x8c @jingyi2811 #512
- [x] OP_2MUL = 0x8d @jingyi2811 #512
- [x] OP_2DIV = 0x8e @jingyi2811 #512
- [x] OP_NEGATE = 0x8f @jingyi2811 #512
- [x] OP_ABS = 0x90 @jingyi2811 #512
- [x] OP_NOT = 0x91 @jingyi2811 #512
- [x] OP_0NOTEQUAL = 0x92 @jingyi2811 #512
- [x] OP_ADD = 0x93 @jingyi2811 #512
- [x] OP_SUB = 0x94 @jingyi2811 #512
- [x] OP_MUL = 0x95 @jingyi2811 #512
- [x] OP_DIV = 0x96 @jingyi2811 #512
- [x] OP_MOD = 0x97 @jingyi2811 #512
- [x] OP_LSHIFT = 0x98 @jingyi2811 #512
- [x] OP_RSHIFT = 0x99 @jingyi2811 #512
- [x] OP_BOOLAND = 0x9a @jingyi2811 #512
- [x] OP_BOOLOR = 0x9b @jingyi2811 #512
- [x] OP_NUMEQUAL = 0x9c @jingyi2811 #512
- [x] OP_NUMEQUALVERIFY = 0x9d @jingyi2811 #512
- [x] OP_NUMNOTEQUAL = 0x9e @jingyi2811 #512
- [x] OP_LESSTHAN = 0x9f @jingyi2811 #512
- [x] OP_GREATERTHAN = 0xa0 @jingyi2811 #512
- [x] OP_LESSTHANOREQUAL = 0xa1 @jingyi2811 #512
- [x] OP_GREATERTHANOREQUAL = 0xa2 @jingyi2811 #512
- [x] OP_MIN = 0xa3 @jingyi2811 #512
- [x] OP_MAX = 0xa4 @jingyi2811 #512
- [x] OP_WITHIN = 0xa5 @jingyi2811 #512

## `crypto.ts`
- [x] OP_RIPEMD160 @jingyi2811 #203
- [x] OP_SHA1 @jingyi2811 #219
- [x] OP_SHA256 @jingyi2811 #219
- [x] OP_HASH160 @fuxingloh #106
- [x] OP_HASH256 @jingyi2811 #219
- [x] OP_CODESEPARATOR  @jingyi2811 #219
- [x] OP_CHECKSIG @fuxingloh #106
- [x] OP_CHECKSIGVERIFY @jingyi2811 #219
- [x] OP_CHECKMULTISIG  @jingyi2811 #219
- [x] OP_CHECKMULTISIGVERIFY @jingyi2811 #219

## `expansion.ts`
- [x] OP_NOP1 = 0xb0 @Jouzo #440
- [x] OP_CHECKLOCKTIMEVERIFY = 0xb1 @Jouzo #440
- [x] OP_NOP2 = OP_CHECKLOCKTIMEVERIFY @Jouzo #440
- [x] OP_CHECKSEQUENCEVERIFY = 0xb2 @Jouzo #440
- [x] OP_NOP3 = OP_CHECKSEQUENCEVERIFY @Jouzo #440
- [x] OP_NOP4 = 0xb3 @Jouzo #440
- [x] OP_NOP5 = 0xb4 @Jouzo #440
- [x] OP_NOP6 = 0xb5 @Jouzo #440
- [x] OP_NOP7 = 0xb6 @Jouzo #440
- [x] OP_NOP8 = 0xb7 @Jouzo #440
- [x] OP_NOP9 = 0xb8 @Jouzo #440
- [x] OP_NOP10 = 0xb9 @Jouzo #440

## `invalid.ts`
- [x] OP_INVALIDOPCODE = 0xff @Jouzo #440

### PR DESCRIPTION
**Pull request by aikchun**
#### What kind of PR is this?:

/kind feature

#### Which issue(s) does this PR fixes?:

Fixes part of #107

### CODE REVIEW

File: packages/jellyfish-transaction/__tests__/script/index.test.ts [hunk 1]
```diff
@@ -720,6 +720,348 @@ describe('[OP_CHECKMULTISIGVERIFY]', () => {
   })
 })
 
+describe('[OP_TOALTSTACK]', () => {
```
Comment (jingyi2811) [inline packages/jellyfish-transaction/__tests__/script/index.test.ts:723]: Move all the code to the line after OP_RETURN.
The hex for OP_RETURN is '016a'
So 016a, 016b, 016c ...

File: packages/jellyfish-transaction/__tests__/script/index.test.ts [hunk 2]
```diff
@@ -720,6 +720,348 @@ describe('[OP_CHECKMULTISIGVERIFY]', () => {
...
+  })
+
+  it('should map toBuffer', () => {
+    const smartBuffer = new SmartBuffer()
+    OP_CODES.toBuffer([OP_CODES.OP_DROP], smartBuffer)
+    expect(smartBuffer.toBuffer().toString('hex')).toStrictEqual(hex)
+  })
+})
+
+describe('[OP_NIP]', () => {
```
Comment (jingyi2811) [inline packages/jellyfish-transaction/__tests__/script/index.test.ts:939]: Move Fuxing's OP_DUP code (Hex 0176) to line 938

File: packages/jellyfish-transaction/__tests__/script/index.test.ts [hunk 3]
```diff
@@ -720,6 +720,348 @@ describe('[OP_CHECKMULTISIGVERIFY]', () => {
...
+  })
+
+  it('should map toBuffer', () => {
+    const smartBuffer = new SmartBuffer()
+    OP_CODES.toBuffer([OP_CODES.OP_2DUP], smartBuffer)
+    expect(smartBuffer.toBuffer().toString('hex')).toStrictEqual(hex)
+  })
+})
+
+describe('[OP_2DUP]', () => {
```
Comment (jingyi2811) [inline packages/jellyfish-transaction/__tests__/script/index.test.ts:795]: You wrote this OP_2DUP describe two times

**Comment by jingyi2811** [comment]
Kindly delete this comment too. 
https://github.com/DeFiCh/jellyfish/blob/0ca1fddad41ee59af372f35bd2db8bce06406555/packages/jellyfish-transaction/src/script/stack.ts#L1

File: packages/jellyfish-transaction/__tests__/script/index.test.ts (hunk 2 above)
Comment (aikchun) [reply packages/jellyfish-transaction/__tests__/script/index.test.ts:939]: Fuxing's `OP_DUP` test code is not a standalone. It doesn't have a standalone `OP_DUP`. Are you talking about all the tests that involves `OP_DUP`

Comment (jingyi2811) [reply packages/jellyfish-transaction/__tests__/script/index.test.ts]: Yup

Comment (jingyi2811) [reply packages/jellyfish-transaction/__tests__/script/index.test.ts]: Help Fuxing to add this 

```
describe('[OP_DUP]', () => {
})
```

**Review by canonbrother — APPROVED** [review]

File: packages/jellyfish-transaction/__tests__/script/mapping.test.ts [hunk 4]
```diff
@@ -138,10 +138,60 @@ describe('All mapped OP_CODES are setup properly: (static, hex, num, asm)', () =
...
 
+  it('OP_TOALTSTACK', () => {
+    expectOPCode(script.OP_CODES.OP_TOALTSTACK, script.OP_TOALTSTACK, 'OP_TOALTSTACK', 0x6b, '6b')
+  })
+
+  it('OP_FROMALTSTACK', () => { expectOPCode(script.OP_CODES.OP_FROMALTSTACK, script.OP_FROMALTSTACK, 'OP_FROMALTSTACK', 0x6c, '6c') })
+
+  it('OP_2DROP', () => { expectOPCode(script.OP_CODES.OP_2DROP, script.OP_2DROP, 'OP_2DROP', 0x6d, '6d') })
+
+  it('OP_3DUP', () => { expectOPCode(script.OP_CODES.OP_3DUP, script.OP_3DUP, 'OP_3DUP', 0x6f, '6f') })
```
Comment (jingyi2811) [inline packages/jellyfish-transaction/__tests__/script/mapping.test.ts:149]: Help Fuxing to add `it` for OP_2DUP

Comment (aikchun) [reply packages/jellyfish-transaction/__tests__/script/mapping.test.ts:151]: Added test for `OP_2DIP`

**Review by jingyi2811 — APPROVED** [review]

**Review by fuxingloh — APPROVED** [review]

**Pull request merged**